apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs

```diff
@@ -362,6 +354,54 @@ public override async Task ModifyAckDeadlineAsync(SubscriptionName subscription,
                     ids.AddRange(ackIds.Select(id => new TimedId(_clock.GetCurrentDateTimeUtc(), id)));
                 }
             }
+
+            /// <summary>
+            /// This method does the following:
+            /// <list type="bullet">
+            /// <item> Checks if subscription under test is exactly once or not.</item>
+            /// <item> If it's not exactly once, it will always throw the specified exception for the specified number of times as <c>_numberOfAckModifyAckDeadlineFailures</c>.</item>
+            /// <item> If it's exactly once, it first checks if the entire request has failed permanently based on <c>StatusCode</c>. If yes, it will set the flag to throw the specified exception.</item>
+            /// <item> If the entire request has not failed permanently, it will check the error metadata to see if it contains permanent error for the given ackId. If yes, it will set the flag to throw the specified exception.</item>
+            /// <item> If the flag to throw the specified exception is set and the specified number of exceptions are not thrown already, the specified exception is thrown, else exception is not thrown.</item>
+            /// </list>
+            /// </summary>
+            /// <param name="ackIds">The list of acknowledgement ids being processed.</param>
+            private void MayBeThrowException(IEnumerable<string> ackIds)
+            {
+                // For non-exactly once subscriptions, we always throw exception.
```

Comment:
Rather than keeping a local variable for this, I would target the *one* situation where you don't want to do anything (exactly-once delivery, an RpcException that is retriable and none of the ack IDs has a permanent error), and return at that point. I think that would be clearer.

In particular, it would make it more obvious that in this situation we *don't* modify `_numberOfAckModifyAckDeadlineFailures` - I'm assuming that was deliberate. If you *did* intend to call `CompareExchange` (and maybe increment `_numberOfAckModifyAckDeadlineFailures`) unconditionally, I'd do that at the start of the method.

Basically at the moment it's not clear to me whether the expected behavior is "modify state and maybe throw an exception" or "maybe ignore completely, but otherwise modify state and throw an exception".

Reply:
I have added the retryable scenario without permanent error as suggested by you. I have also updated the comments to make the intention of the method clear. Please let me know if further changes are needed and I am happy to do so.

Reply:
It still feels to me like there's a lot of code (30 lines) which is basically "should we immediately return" - but it's not written that way. (And in fact at the moment in one case you *are* returning, and in another you're just setting `throwException` to false, when they're equivalent - and that's confusing.)

Reply:
Sure Jon. Understood. I will try to refactor it in next commit and we can take it further from there.